File: web/src/entities/player-data.ts
import { z } from "zod";

const playerDataSchema = z.object({
  nickname: z.string(),
  totalTime: z.number(),
  augmentIds: z.array(z.number()),
  logs: z.array(
    z.object({
      time: z.number(),
      x: z.number(),
      y: z.number(),
      z: z.number(),
    }),
  ),
});

export type PlayerData = z.infer<typeof playerDataSchema>;
export const parsePlayerData = playerDataSchema.parse;

File: SchoolRush/Assets/Scripts/Data/PlayerData.cs
using UnityEngine;
using System.Collections.Generic;

[System.Serializable]
public class PlayerData {
    public string nickname;
    public int totalTime;
    [SerializeField] public List<int> augmentIds;
    [SerializeField] public List<Log> logs;

    public PlayerData(string nickname) {
        this.nickname = nickname;
        augmentIds = new List<int>();
        logs = new List<Log>();
    }

    public void Insert(Vector3 position) {
        this.logs.Add(new Log((int)System.DateTimeOffset.UtcNow.ToUnixTimeSeconds(), position));
    }

    public void Save(int totalTime) {
        this.totalTime = totalTime;
        PlayerDataSaveLoad.SaveData(this);
    }
}

[System.Serializable]
public class Log {
    public int time;
    public float x, y, z;

    public Log (int time, Vector3 position) {
      this.time = time;
      this.x = position.x;
      this.y = position.y;
      this.z = position.z;
    }
}
